src/Sarif/Baseline/V2/TrustMap.cs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Baseline
+{
+    /// <summary>
+    ///  TrustMap determines how much to trust different Result attributes for matching.
+    ///  
+    ///  Trust is the product of how often values match and how unique the values were.
+    ///  An attribute which always had the same value will have very low trust.
+    ///  An attribute which never matches will have very low trust.
+    /// </summary>
+    internal class TrustMap
+    {
+        public const float DefaultTrust = 0.1f;
+        public bool WasMatched { get; private set; }
+        private Dictionary<TrustKey, TrustValue> Map { get; }
+
+        public TrustMap()
+        {
+            Map = new Dictionary<TrustKey, TrustValue>();
+        }
+
+        /// <summary>
+        ///  Return the matching trust of a given property.
+        /// </summary>
+        /// <param name="propertySetName">Set name (Base, PartialFingerprints, Properties, etc)</param>
+        /// <param name="propertyName">Property Name (valueFingerprint/v1)</param>
+        /// <returns>Value between zero and one indicating how much to trust this property for matching</returns>
+        public float Trust(string propertySetName, string propertyName)
+        {
+            if (Map.TryGetValue(new TrustKey(propertySetName, propertyName), out TrustValue value))
+            {
+                // No trust for constant values - more than one occurrence but only one value
+                if (value.UseCount > 1 && value.UniqueValues.Count == 1) { return 0; }
+
+                // Unique values but no matches defaults to low trust; no matches can occur with useful attributes
+
+                if (WasMatched)
+                {
+                    // Simplification of (Match / Unique) * (Unique / Use)
+                    // The product of how many values match and how many unique values there were
+                    return Math.Max(DefaultTrust, ((float)value.MatchCount) / (float)(value.UseCount));
+                }
+                else
+                {
+                    return Math.Max(DefaultTrust, ((float)value.UniqueValues.Count) / (float)(value.UseCount));
+                }
+            }
+
+            // Unknown
+            return DefaultTrust;
+        }
+
+        /// <summary>
+        ///  Add a Result attribute to the map.
+        /// </summary>
+        /// <param name="component">WhatComponent for a Result attribute used in baselining</param>
+        public void Add(WhatComponent component)
+        {
+            TrustKey key = new TrustKey(component);
+
+            TrustValue value = null;
```

Comment:
> = null [](start = 28, length = 7)

Nit: `= null` unnecessary because `out` parameter is guaranteed to be assigned. But even better, you can now declare the `out` parameter inline:
```
if (!Map.TryGetValue(key, out TrustValue value))
```

Reply:
Yes. I sometimes have compound conditions and then the method doesn't always fire and the compiler requires the default assignment. I like to leave it there just to be clear to the reader what value it will have if for some reason the call doesn't happen.

The sneakiest are conditions like this:
TrustValue value = null;
if (!Map?.TryGetValue(key, out value)) {
 ...
}

Here, thanks to ?., if Map is null, the method will not get called at all.